google/cloud/internal/async_read_write_stream_auth.h

```diff
@@ -28,79 +29,97 @@ namespace internal {
       std::unique_ptr<grpc::ClientContext> context,
       std::shared_ptr<GrpcAuthenticationStrategy> auth, StreamFactory factory)
-      : context_(std::move(context)),
-        auth_(std::move(auth)),
-        factory_(std::move(factory)) {}
+      : auth_(std::move(auth)),
+        state_(std::make_shared<SharedState>(std::move(factory), std::move(context))) {}
 
   void Cancel() override {
-    if (context_) return context_->TryCancel();
-    if (stream_) return stream_->Cancel();
+    state_->Cancel();
   }
 
   future<bool> Start() override {
     using Result = StatusOr<std::unique_ptr<grpc::ClientContext>>;
 
-    auto weak =
-        std::weak_ptr<AsyncStreamingReadWriteRpcAuth>(this->shared_from_this());
-    return auth_->AsyncConfigureContext(std::move(context_))
+    std::unique_ptr<grpc::ClientContext> context;
+    {
+      std::lock_guard<std::mutex> g{state_->mu};
+      context = std::move(state_->initial_context);
+    }
+    auto weak = std::weak_ptr<SharedState>(state_);
+    return auth_->AsyncConfigureContext(std::move(context))
         .then([weak](future<Result> f) mutable {
-          if (auto self = weak.lock()) return self->OnStart(f.get());
+          if (auto state = weak.lock()) return state->OnStart(f.get());
           return make_ready_future(false);
         });
   }
 
   future<absl::optional<Response>> Read() override {
-    if (!stream_) return make_ready_future(absl::optional<Response>{});
-    return stream_->Read();
+    std::lock_guard<std::mutex> g{state_->mu};
+    return state_->stream->Read();
   }
 
   future<bool> Write(Request const& request,
                      grpc::WriteOptions options) override {
-    if (!stream_) return make_ready_future(false);
-    return stream_->Write(request, std::move(options));
+    std::lock_guard<std::mutex> g{state_->mu};
+    return state_->stream->Write(request, std::move(options));
   }
 
   future<bool> WritesDone() override {
-    if (!stream_) return make_ready_future(false);
-    return stream_->WritesDone();
+    std::lock_guard<std::mutex> g{state_->mu};
+    return state_->stream->WritesDone();
   }
 
   future<Status> Finish() override {
-    if (!stream_) {
-      return make_ready_future(
-          Status(StatusCode::kInvalidArgument,
-                 "uninitialized GrpcReadWriteStreamAuth<>"));
-    }
-    return stream_->Finish();
+    return state_->Finish();
   }
 
  private:
-  future<bool> OnStart(StatusOr<std::unique_ptr<grpc::ClientContext>> context) {
-    if (!context) {
-      stream_ =
-          absl::make_unique<AsyncStreamingReadWriteRpcError<Request, Response>>(
+  struct SharedState {
+    SharedState(StreamFactory factory, std::unique_ptr<grpc::ClientContext> initial_context) : factory(std::move(factory)), initial_context(std::move(initial_context)), stream(absl::make_unique<AsyncStreamingReadWriteRpcError<Request, Response>>(Status(StatusCode::kFailedPrecondition, "Stream is not yet started."))) {}
```

Comment:
I am sure the robots are unhappy about this....

Reply:
I wonder if `kInternal` is better than `kFailedPrecondition` in this case.  After all, using this initial stream would be a bug in the implementation of the client library?  Thoughts?

Reply:
Changed to kInternal, I don't have a strong preference.